### ISSUE TITLE
[react-native] Wrong type for component ref

### ISSUE THREAD

**Issue by jemise111** [DefinitelyTyped/DefinitelyTyped#16318]
```
class EditProfile extends React.Component<Props, State> {
  _textInputRef: TextInputStatic;

  render() {
    <TextInput ref={ref => this._textInputRef = ref} />
  }
}
```

What is the appropriate type to use in for `_textInputRef` in this situation. In the example above the compiler complains that 

```
Type 'Component<TextInputProperties, ComponentState>' is not assignable to type 'TextInputStatic'.
[0]   Property 'State' is missing in type 'Component<TextInputProperties, ComponentState>'.
```

But if I change the type to `Component<TextInputProperties, ComponentState>` I can no longer call `.focus()` on my input ref.

Is this a bug or am I using the wrong type altogether? Thanks for any help!

(this also applies to other components, not just TextInputs)

I'm using Type definitions for react-native 0.43 (v0.43.9).

- [x] I tried using the `@types/react-native` package and had problems.
- [x] I tried using the latest stable version of tsc. https://www.npmjs.com/package/typescript
- [x] I have a question that is inappropriate for [StackOverflow](https://stackoverflow.com/).  (Please ask any appropriate questions there).
- [x] [Mention](https://github.com/blog/821-mention-somebody-they-re-notified) the authors (see `Definitions by:` in `index.d.ts`) so they can respond.
  - Authors: @....

**Comment by andy-ms** [comment]
You forgot to mention the authors. @alloy @gyzerok

**Comment by alloy** [comment]
Hmm, yeah the `focus()` function is [defined on `NativeMethodsMixinStatic`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/d0dab167cda6b9c39da7280f8b988d3ad8632b68/types/react-native/index.d.ts#L327) and is [extended by `TextInputStatic`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/d0dab167cda6b9c39da7280f8b988d3ad8632b68/types/react-native/index.d.ts#L1237).

Do you mean that this worked with previous typings but no longer after upgrading?

**Comment by gyzerok** [comment]
@jemise111 thank you for your report!

@alloy actually I had the same issue with `Navigator` at some point, but totally forgot it. Just came back from vacation, may take a look myself during the week.

**Comment by jemise111** [comment]
@andy-ms Whoops, thanks!

@alloy To be honest we didn't have a strong type on our `refs`. I took the opportunity to add them after upgrading to 43 and am now encountering this issue.

@gyzerok @alloy  Thank you for taking the time!

**Comment by jemise111** [comment]
@davidfant Unfortunately no, haven't had time to dig into this so I have a few components casted to `any` 😳

**Comment by gyzerok** [comment]
Hi folks!

Sorry, haven't had time to look into it. Feel free to dig into the types and come up with a PR :)

**Comment by bherila** [comment]
@gyzerok @davidfant @jemise111 Any update on this? Happy to send a PR if someone can advise how to fix it.

**Comment by tkrotoff** [comment]
Same issue here.
Using React definition you can write `<input ref={input => this.input = input} />` with input being `HTMLInputElement | null` (btw why `null`?)

@jemise111 `_textInputRef: TextInputStatic` should be simply `_textInputRef: TextInput`, TextInputStatic is kind of an internal stuff I think.

For now I perform a `any` cast: `ref={input => this.username = input as any}`, [example](https://github.com/tkrotoff/react-form-with-constraints/blob/f0234b4001821bb6b195f79dc0618c5cf860bed5/examples/ReactNative/App.tsx#L95):

```TypeScript
class Form extends React.Component<Props, State> {
  username: TextInput;

  render() {
    return (
      <View>
        <Text onPress={() => this.username.focus()}>Username</Text>
        <TextInput
          ref={input => this.username = input as any}
          value={this.state.username}
          onChangeText={this.handleUsernameChange}
        />
      </View>
    );
  }
}
```

**Comment by vkrol** [comment]
@tkrotoff 
> btw why null?

> React will call the ref callback with the DOM element when the component mounts, and **call it with null when it unmounts**.
https://reactjs.org/docs/refs-and-the-dom.html#adding-a-ref-to-a-dom-element

**Comment by danilobuerger** [comment]
any news on this?

**Comment by dawnmist** [comment]
I found a way to make the existing definitions work.
```javascript
import * as React from 'react';
import { NativeMethodsMixin, TextInput, TextInputProperties, TimerMixin } from 'react-native';

type TextInputField = 
  & React.Component<TextInputProperties, React.ComponentState>
  & NativeMethodsMixin
  & TimerMixin;

interface FormProps {
  isValid(value: string): boolean;
  submit(value: string): void;
}

interface FormState {
  value: string;
  error?: string;
}

export class MyForm extends React.Component<FormProps, FormState> {
  field: TextInputField | null = null;

  constructor(props: FormProps) {
    super(props);
    this.state = { value: '' };
  }

  render() {
    return (
      <TextInput
        ref={(node: TextInputField) => {
          // Note: it is required to set the node type to include the
          // Mixins - otherwise it defaults to the component part of
          // the static definition only.
          this.field = node;
        }}
        value={this.state.value}
        onChangeText={(value: string) => this.setState({ value })}
        onSubmitEditing={this.validateAndSubmit}
      />
    );
  }

  private validateAndSubmit = () => {
    const { value } = this.state;
    if (!this.props.isValid(value)) {
      this.setState({ error: `${value} is not valid` }, () => {
        if (this.field !== null) {
          // Return focus to the field if invalid
          this.field.focus();
        }
      });
    }
    else {
      this.props.submit(this.state.value);
      this.setState({ error: undefined });
    }
  };
}
```

What I believe is happening is that the default ref function is only aware of the component portion of the TextInputStatic's definition, so the ref function's input gets defined as `React.Component<TextInputProperties, React.ComponentState>` only. This then means that as far as TypeScript is aware the Mixin methods are "sliced off" the item that is passed into the ref function, so that they cannot be accessed from the ref.

By defining the ref function's input parameter to contain the full definition for TextInput, it tells the TypeScript compiler that the ref's item also contains the functions defined in the Mixins.

**Comment by tkrotoff** [comment]
tl;dr [React Native index.d.ts](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/f929c81119f6103fe7e2ce8ee16102a3d6d832ee/types/react-native/index.d.ts) solution:

```TypeScript
type Constructor<T> = new(...args: any[]) => T;
declare class TextInputComponent extends React.Component<TextInputProperties, TextInputState> {}
declare const TextInputBase: Constructor<NativeMethodsMixin> & Constructor<TimerMixin> & typeof TextInputComponent;
export class TextInput extends TextInputBase {
    State: TextInputState;

    /**
     * Returns if the input is currently focused.
     */
    isFocused: () => boolean;

    /**
     * Removes all text from the input.
     */
    clear: () => void;
}
```

**Waiting for suggestions to improve the code**

Example to check it works (to include at the bottom of [react-native/test/index.tsx](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/f929c81119f6103fe7e2ce8ee16102a3d6d832ee/types/react-native/test/index.tsx)):
```TypeScript
class FormWithTextInputRef extends React.Component<{}, {username: string}> {
    username: TextInput | null = null;

    handleUsernameChange() {
    }

    render() {
        return (
            <View>
                <Text onPress={() => this.username.focus()}>Username</Text>
                <TextInput
                    ref={input => this.username = input}
                    value={this.state.username}
                    onChangeText={this.handleUsernameChange}
                />
            </View>
        );
    }
}
```

<br>
<br>
<br>
<br>
<br>
<br>

Explanations
------

In React DOM, you have 2 separated things: `<input />` and `HTMLInputElement` and you write things like that:

```TypeScript
class Form extends React.Component<Props, State> {
  username: HTMLInputElement | null = null;

  render() {
    return (
      <div>
        <input
          ref={input => this.username = input}
          value={this.state.username}
          onChange={this.handleUsernameChange}
        />
      </div>
    );
  }
}
```

In React Native, there is no separation between the markup (`<input />` vs `<TextInput />`) and the instance (`HTMLInputElement` vs `TextInput`).

The solution would be to write a real class inside `react-native/index.d.ts`:
```TypeScript
export class TextInput extends React.Component<TextInputProps, TextInputState> {
  // ...
}
```

Problem, [current `index.d.ts` is full of mixins](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/f929c81119f6103fe7e2ce8ee16102a3d6d832ee/types/react-native/index.d.ts#L1282) and you can't have multiple inheritance with TypeScript class:

```TypeScript
export interface TextInputStatic extends NativeMethodsMixin, TimerMixin, React.ComponentClass<TextInputProperties> {
    State: TextInputState;

    /**
     * Returns if the input is currently focused.
     */
    isFocused: () => boolean;

    /**
     * Removes all text from the input.
     */
    clear: () => void;
}
```

So I've checked how TypeScript generates a d.ts file from [my own code that contains mixins](https://github.com/tkrotoff/react-form-with-constraints/blob/v0.7.1/packages/react-form-with-constraints/src/FormWithConstraints.tsx#L67) and voila.

No need anymore for [`var TextInput`/`type TextInput`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/f929c81119f6103fe7e2ce8ee16102a3d6d832ee/types/react-native/index.d.ts#L8522):
```TypeScript
export var TextInput: TextInputStatic;
export type TextInput = TextInputStatic;
```

**Comment by methodbox** [comment]
Just a note that I've had effectively the exact same issue using the `Camera` component of Expo.

I'm going to try to fanagle something like @dawnmist 's solution and see if that does the trick.

The problem I ran into is that ref is used like this:

```js
ref={ref => { this.camera = ref; }}
```

...and of course `camera` (lowercase intentional) has not be defined on the component; it refers to the instance of `Camera` which doesn't exist yet; it poses an oddly existential question about applying types to non-instantiated objects.

I'm not sure this PR is going to cover the full ground needed to actually "fix" this.

**Comment by orta** [comment]
Hi thread, we're moving DefinitelyTyped to use [GitHub Discussions](https://github.com/DefinitelyTyped/DefinitelyTyped/issues/53377) for conversations the `@types` modules in DefinitelyTyped.

To help with the transition, we're closing all issues which haven't had activity in the last 6 months, which includes this issue. If you think closing this issue is a mistake, please pop into the [TypeScript Community Discord](https://discord.gg/typescript) and mention the issue in the `definitely-typed` channel.
